### PR TITLE
Bundle diff output with charm url names

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following allows the diff'ing of bundles when using charmhub charm
urls. The internal representation of charmhub urls are different from
what users of juju see. So prefix, arch, series charm urls should not be
shown and therefor should not be used for diff'ing exported bundles.

The code is actually simple, use the charmurl name for charmhub charms
and just update to the latest library changes. With the addition of a
few tests, the following should be possible:

    $ juju diff-bundle <(juju export-bundle)
    {}

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy ubuntu --channel="latest/edge"
$ juju diff-bundle <(juju export-bundle)
{}
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1913631

### CODE REVIEW

**Comment by aieri** [comment]
> With the addition of a
> few tests, the following should be possible:
> 
> ```
> $ juju diff-bundle <(juju export-bundle)
> {}
> ```

Kind of a sidenote, but ```juju diff-bundle <(juju export-bundle)``` is already not providing the expected result due to https://bugs.launchpad.net/juju/+bug/1913631 (independently from charmhub url issues)

**Comment by SimonRichardson** [comment]
@aieri that should be fixed here https://github.com/juju/juju/pull/12638

**Comment by hmlanigan** [comment]
From QA: export bundle is exporting the risk only, not the track/risk combo where needed.  

$ juju deploy ubuntu-juju-qa --channel 2.0/stable
Located charm "ubuntu-juju-qa" in charm-hub, revision 12
Deploying "ubuntu-juju-qa" from charm-hub charm "ubuntu-juju-qa", revision 12 in channel 2.0/stable
$ juju export-bundle
...
  ubuntu-juju-qa:
    charm: ubuntu-juju-qa
    channel: stable
    series: bionic
...

File: cmd/juju/application/diffbundle_test.go [hunk 1]
```diff
@@ -776,5 +794,34 @@ applications:
...
+    charm: 'cs:prometheus2-7'
+    num_units: 1
+    series: xenial
+    constraints: 'cores=3'
+    options:
+      ontology: kant
+    to:
+      - 0
+  grafana:
+    charm: 'cs:grafana-19'
```
Comment (hmlanigan) [inline cmd/juju/application/diffbundle_test.go:811]: Can we have one charm CS and one CH in the test?

**Comment by hmlanigan** [comment]
Partial fix for https://bugs.launchpad.net/juju/+bug/1915844 also

**Review by hmlanigan — APPROVED** [review]
QA on charm piece good.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**